parsing/pcoq.ml

```diff
@@ -571,14 +590,15 @@ let factorize_grams l1 l2 =
 let number_of_entries gcl =
   List.fold_left (fun n (p,_,_) -> n + p) 0 gcl
 
-let unfreeze (grams, lex, names) =
+let unfreeze (grams, lex, names, custom) =
   let (undo, redo, common) = factorize_grams !grammar_stack grams in
   let n = number_of_entries undo in
   remove_grammars n;
   grammar_stack := common;
   CLexer.set_keyword_state lex;
   List.iter extend_dyn_grammar (List.rev_map pi2 redo);
-  grammar_names := names
+  grammar_names := names;
+  custom_entries := custom
 
```

Comment:
I don't think this is enough, because grammar entries are mutable objects. So you will still encounter backtracking problems with this, you probably need to store more information about the transient entries in the grammar stack.

Reply:
Once rules of a grammar entry have been deleted (and backtrack shall do that for `Notation`, right?), aren't the grammar entries empty at the time we undo the `Declare Custom` for the corresponding entry, and thus ineffective from the user point of view?

Reply:
Good question. This would need testing to check whether it is the case. I don't have a constructive proof, but I am confident that you can find race conditions though if you do not undo/redo linearly, as it is the case in some uses of the STM.

Furthermore, this code will likely fail when marshalling data through processes as grammars contain closure in general. I'd thus really advise against storing them in the summary. You need to implement some handler system instead, which should be immutable and marshallable data.

Reply:
What @ppedrot means is not that complicated, whatever state you put in the summary [it should be the full state I'm afraid], make sure it can be marshalled. That is, no closures.

Reply:
Both the `custom_entries` and `grammar_names` are camlp5 grammar entries, so I guess they contain closures.

Reply:
`grammar_names` is not part of the summary IIUC.

Reply:
It needs to be as soon as custom entries are registered into it.